### PR TITLE
UCT/IB/RC: Fence PCI atomics on verbs RC

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9181/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9182/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9281/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9308/ for details (Mellanox internal link).

File: src/uct/ib/rc/base/rc_iface.c [hunk 1]
```diff
@@ -150,39 +150,52 @@ ucs_status_t uct_rc_iface_query(uct_rc_iface_t *iface,
...
+        }
+        if (dev->pci_cswap_arg_sizes & sizeof(uint32_t)) {
+            iface_attr->cap.atomic32.fop_flags |= UCS_BIT(UCT_ATOMIC_OP_CSWAP);
+        }
+#endif
+        iface_attr->cap.flags                  |= UCT_IFACE_FLAG_ATOMIC_CPU;
+    } else {
+        if (dev->atomic_arg_sizes & sizeof(uint64_t)) {
+            /* TODO: remove deprecated flags */
+            iface_attr->cap.flags              |= UCT_IFACE_FLAG_ATOMIC_DEVICE;
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_iface.c]: why is this inside 'if'? should be right before line 172, in the 'else' clause of `if (dev->pci_fadd_arg_sizes || dev->pci_cswap_arg_sizes) {`

File: src/uct/ib/rc/verbs/rc_verbs.h [hunk 2]
```diff
@@ -32,6 +32,10 @@ typedef struct uct_rc_verbs_txcnt {
 typedef struct uct_rc_verbs_ep {
     uct_rc_ep_t            super;
     uct_rc_verbs_txcnt_t   txcnt;
+    struct {
+        uint16_t                fence_beat;
```
Comment (yosefe) [inline src/uct/ib/rc/verbs/rc_verbs.h]: indent on column.
can we extract common code with mlx5?

File: src/uct/ib/rc/verbs/rc_verbs_ep.c [hunk 3]
```diff
@@ -90,6 +90,21 @@ uct_rc_verbs_ep_post_send_desc(uct_rc_verbs_ep_t* ep, struct ibv_send_wr *wr,
     uct_rc_txqp_add_send_op_sn(&ep->super.txqp, &desc->super, ep->txcnt.pi);
 }
 
+static UCS_F_ALWAYS_INLINE int
+uct_rc_verbs_ep_atomic_fence(uct_rc_verbs_iface_t *iface,
```
Comment (yosefe) [inline src/uct/ib/rc/verbs/rc_verbs_ep.c]: need to reuse code with mlx5

File: src/uct/ib/rc/verbs/rc_verbs_ep.c [hunk 4]
```diff
@@ -100,14 +115,18 @@ uct_rc_verbs_ep_rdma_zcopy(uct_rc_verbs_ep_t *ep, const uct_iov_t *iov,
...
     size_t sge_cnt;
+    int send_flags = IBV_SEND_SIGNALED;
 
     UCT_RC_CHECK_RES(&iface->super, &ep->super);
     sge_cnt = uct_ib_verbs_sge_fill_iov(sge, iov, iovcnt);
     UCT_SKIP_ZERO_LENGTH(sge_cnt);
     UCT_RC_VERBS_FILL_RDMA_WR_IOV(wr, wr.opcode, opcode, sge, sge_cnt, remote_addr, rkey);
     wr.next = NULL;
+    if (opcode == IBV_WR_RDMA_READ) {
+        send_flags |= uct_rc_verbs_ep_atomic_fence(iface, ep, IBV_SEND_FENCE);
```
Comment (yosefe) [inline src/uct/ib/rc/verbs/rc_verbs_ep.c]: what about get_bcopy?
maybe we can add it to common code, uct_rc_verbs_ep_post_send()?

File: src/uct/ib/rc/verbs/rc_verbs_iface.c [hunk 5]
```diff
@@ -198,6 +202,7 @@ static UCS_CLASS_INIT_FUNC(uct_rc_verbs_iface_t, uct_md_h md, uct_worker_h worke
                                                self->super.config.tx_qp_len);
     self->super.config.tx_moderation = ucs_min(self->super.config.tx_moderation,
                                                self->config.tx_max_wr / 4);
+    self->config.no_fence            = !config->fence;
```
Comment (yosefe) [inline src/uct/ib/rc/verbs/rc_verbs_iface.c]: why inverse the meaning of the "fence" flag when it's copied to iface? IMHO it's more confusing
let's have just 'self->config.fence'

File: src/uct/ib/rc/base/rc_iface.c [hunk 6]
```diff
@@ -150,39 +150,52 @@ ucs_status_t uct_rc_iface_query(uct_rc_iface_t *iface,
...
+#if HAVE_IB_EXT_ATOMICS
+        if (dev->pci_fadd_arg_sizes & sizeof(uint32_t)) {
+            iface_attr->cap.atomic32.op_flags  |= UCS_BIT(UCT_ATOMIC_OP_ADD);
+            iface_attr->cap.atomic32.fop_flags |= UCS_BIT(UCT_ATOMIC_OP_ADD);
+        }
+        if (dev->pci_cswap_arg_sizes & sizeof(uint32_t)) {
+            iface_attr->cap.atomic32.fop_flags |= UCS_BIT(UCT_ATOMIC_OP_CSWAP);
+        }
+#endif
+        iface_attr->cap.flags                  |= UCT_IFACE_FLAG_ATOMIC_CPU;
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_iface.c]: push this inside each of the 'if' -s above

Comment (Artemy-Mellanox) [reply src/uct/ib/rc/base/rc_iface.c]: We still use (md->dev.pci_fadd_arg_sizes || md->dev.pci_cswap_arg_sizes) as a check for pci atomic in uct_rc_mlx5_iface_fence(). Maybe return UCT_IB_DEVICE_FLAG_PCI_ATOMICS?

Comment (yosefe) [reply src/uct/ib/rc/base/rc_iface.c]: where is this flag defined?

Comment (Artemy-Mellanox) [reply src/uct/ib/rc/base/rc_iface.c]: Was added here https://github.com/openucx/ucx/pull/3226/commits/f5c3cc5cb734c419574c642d777e6847ad1b0751, removed here https://github.com/openucx/ucx/pull/3226/commits/42be17a8d93f996d67dce32187189eb5e13b84e3

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9367/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9368/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9369/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9370/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9372/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
@yosefe

File: src/uct/ib/base/ib_iface.h [hunk 7]
```diff
@@ -200,6 +200,15 @@ struct uct_ib_iface {
     uct_ib_iface_ops_t      *ops;
 };
 
+
+typedef struct uct_fence_info {
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.h]: uct_ib_fence_state

File: src/uct/ib/base/ib_iface.h [hunk 8]
```diff
@@ -553,4 +562,9 @@ size_t uct_ib_iface_hdr_size(size_t max_inline, size_t min_size)
     return (size_t)ucs_max((ssize_t)(max_inline - min_size), 0);
 }
 
+static UCS_F_ALWAYS_INLINE void uct_init_fi(uct_fence_info_t* fi)
+{
+    fi->fence_beat = 0;
+    fi->fence_flag = 0;
+}
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.h]: space line after

File: src/uct/ib/base/ib_iface.h [hunk 9]
```diff
@@ -553,4 +562,9 @@ size_t uct_ib_iface_hdr_size(size_t max_inline, size_t min_size)
     return (size_t)ucs_max((ssize_t)(max_inline - min_size), 0);
 }
 
+static UCS_F_ALWAYS_INLINE void uct_init_fi(uct_fence_info_t* fi)
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.h]: `uct_ib_fence_state_init(uct_fence_state_t* fence)`

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 10]
```diff
@@ -444,6 +444,9 @@ UCS_CLASS_INIT_FUNC(uct_rc_mlx5_iface_common_t,
         goto cleanup_tm;
     }
 
+    self->super.config.fence = !!((dev->pci_fadd_arg_sizes | dev->pci_cswap_arg_sizes) &
+                                  (sizeof(uint32_t) | sizeof(uint64_t)));
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c]: 👍

File: src/uct/ib/rc/base/rc_ep.h [hunk 11]
```diff
@@ -445,4 +445,36 @@ uct_rc_fc_req_moderation(uct_rc_fc_t *fc, uct_rc_iface_t *iface)
...
+    /* in case PCI Atomics are enabled atomic/read operation on target
+     * are unordered according to PCI specification so we need to
+     * request atomic fence for next such operation */
+    if (fence && iface->config.fence) {
+        fi->fence_flag = 1;
+    }
+
+    UCT_TL_EP_STAT_FENCE(ucs_derived_of(tl_ep, uct_base_ep_t));
+    return UCS_OK;
+}
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_ep.h]: space line after

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 12]
```diff
@@ -697,14 +697,15 @@ void uct_rc_mlx5_iface_common_query(uct_ib_iface_t *ib_iface,
     iface_attr->cap.flags        |= UCT_IFACE_FLAG_ERRHANDLE_ZCOPY_BUF |
                                     UCT_IFACE_FLAG_ERRHANDLE_REMOTE_MEM;
 
-    if (dev->pci_fadd_arg_sizes || dev->pci_cswap_arg_sizes) {
+    if (iface->super.config.fence) {
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_common.c]: this doesn't make sense when reading it (even though it's correct)
maybe use same inline function which checks dev->pci_fadd_arg_sizes and dev->pci_cswap_arg_sizes, both here, and when initializing the config.fence

File: src/uct/ib/rc/base/rc_iface.c (hunk 6 above)
Comment (yosefe) [reply src/uct/ib/rc/base/rc_iface.c]: this comment is not addressed

File: src/uct/ib/rc/verbs/rc_verbs_ep.c (hunk 4 above)
Comment (yosefe) [reply src/uct/ib/rc/verbs/rc_verbs_ep.c]: this comment is not addressed

File: src/uct/ib/rc/base/rc_iface.c (hunk 6 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/rc/base/rc_iface.c]: since if (<pci atomics>) is true iff we work in pci atomics mode no need to push inside each "if" now

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9411/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9413/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9444/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9453/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9483/ for details (Mellanox internal link).

**Pull request merged**